keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,928 @@
+
+### Non-Goals
+
+- Change the existing behavior of Jobs without a SuccessPolicy.
+- Support SuccessPolicy for the job with `NonIndexed` mode: The SuccessPolicy can be theoretically supported for the job with `NonIndexed` mode. 
+However, we don't work on the job with `NonIndexed` mode in the first iteration since there aren't any effective use cases for the NonIndexed job.
+
+## Proposal
+
+We propose new policies under which a job can be declared as succeeded.
+Those policies can be modeled in the following:
+
+1. An indexed job completes if a set of [x, y, z...] indexes are successful.
+2. An indexed job completes if x of indexes are successful.
+
+Then, when the job meets one of the success policies, a new condition, `SuccessCriteriaMet,` is added.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a machine-learning researcher, I run an indexed job which a leader is launched
+as index=0 and workers are launched as index=1+.
+I want to care about only the leaders when the result of job is evaluated.
+
+In addition, we want to terminate the lingering pods if the leader index (index=0) is succeeded
+because the workers often don't have any ways to terminate themselves due to launching daemon processes like ssh-server.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+    criteria:
+    - succeededIndexes: "0"
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 2
+
+As a simulation researcher/engineer for fluid dynamics/biochemistry.
+I want to mark the job as successful and terminate the lingering pods
+when the job meets the one of following conditions:
+
+1. The case of the leader index (index=0) is succeeded
+2. The case of some worker indexes (index=1+) are succeeded
+
+Because succeeded leader index means that the whole simulation is succeeded,
+and succeeded some worker indexes means that the minimum required value is satisfied.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+    criteria:
+    - succeededIndexes: 0
+    - succeededCount: 5
+      succeededIndexes: "1-9"
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+#### No support JobSuccessPolicy for the NonIndexed Job
+
+As I described in [Non-Goals](#non-goals), we don't support the SuccessPolicy for the job with `NonIndexed` mode.
+
+#### Difference between "complete" and "successCriteriaMet"
+The similar job conditions, `Complete` and `SuccessCriteriaMet`, are different in the following ways:
+
+- `Complete` means that all pods completed, not failed.
```

Comment:
```suggestion
- `Complete` means that, if defined, the success criteria was satisfied and all pods are now completed.
```

Reply:
@alculquicondor  When the successPolicy isn't specified, Complete doesn't mean the success criteria were satisfied.

So, I think that the current definition would be better.

Reply:
How about this?
```suggestion
- `Complete` means that all pods completed and either all of them were successful or the Job already had `SuccessCriteriaMet=true`.
```

Reply:
Done.